src/DataReader/proto_reader.rs

```diff
@@ -0,0 +1,275 @@
+    simulation_info: &SimulationInfo,
+) -> TransitionSystemPtr {
+    let composition = simulation_info.component_composition.to_owned();
+    let component_info = simulation_info.components_info.as_ref().unwrap();
+
+    let mut component_container = ComponentContainer::from_info(component_info).unwrap();
+
+    component_loader_to_transition_system(&mut component_container, &composition)
+}
+
+/// Borrows a [`ComponentsInfo`] and returns the corresponding [`Vec`] of [`Component`]s.
+pub fn components_info_to_components(components_info: &ComponentsInfo) -> Vec<Component> {
+    components_info
+        .components
+        .iter()
+        .flat_map(parse_components_if_some)
+        .flatten()
+        .collect()
+}
+
+/// Consumes a [`ProtoDecision`] and the borrows the [`TransitionsSystemPtr`] it belongs to and returns the corresponding [`Decision`].
+///
+/// # Panics
+/// If:
+/// - `proto_decision.source` is `None`.
+/// - `proto_decision.edge` is `None`.
+pub fn proto_decision_to_decision(
+    proto_decision: ProtoDecision,
+    system: &TransitionSystemPtr,
+    components: Vec<Component>,
+) -> Decision {
+    let proto_state: ProtoState = proto_decision.source.unwrap();
+    let state = proto_state_to_state(proto_state, system);
+
+    let proto_edge: ProtoEdge = proto_decision.edge.unwrap();
+    let decided = proto_edge_to_edge(proto_edge, components);
+
+    Decision::new(state, decided)
+}
+
+/// Consumes a [`ProtoState`] and the borrows the [`TransitionsSystemPtr`] it belongs to and returns the corresponding [`State`].
+///
+/// # Panics
+/// If:
+/// - `state.federation` is `None`.
+/// - `state.location_tuple` is `None`.
+pub fn proto_state_to_state(state: ProtoState, system: &TransitionSystemPtr) -> State {
+    let proto_federation: ProtoFederation = state.federation.unwrap();
+    let federation: OwnedFederation =
+        proto_federation_to_owned_federation(proto_federation, system);
+
+    let proto_location_tuple: ProtoLocationTuple = state.location_tuple.unwrap();
+    let location_tuple =
+        proto_location_tuple_to_location_tuple(&proto_location_tuple, system).unwrap();
+
+    State::create(location_tuple, federation)
+}
+
+fn proto_location_tuple_to_location_tuple(
+    location_tuple: &ProtoLocationTuple,
+    system: &TransitionSystemPtr,
+) -> Option<LocationTuple> {
+    let id_looking_for: Vec<LocationID> = location_tuple
+        .locations
+        .iter()
+        .map(|l| LocationID::Simple {
+            location_id: l.id.to_string(),
+            component_id: l
+                .specific_component
+                .as_ref()
+                .map(|c| c.component_name.to_string()),
+        })
+        .collect();
+
+    system
+        .get_all_locations()
+        .into_iter()
+        .map(|tuple| (tuple.id.clone(), tuple))
+        .map(|(id, tuple)| (id.inorder_vec_tranform(), tuple))
+        .filter(|(id, _)| id.iter().eq(id_looking_for.iter()))
+        .collect::<Vec<_>>()
```

Comment:
Do you need to collect to take the first element?

Reply:
No, `.find()` would simplify this even more.